### PR TITLE
Fix loading of CRLs and certs from a file.

### PR DESCRIPTION
**Pull request by SparkiDev**
Multiple CRLs and Certs parsed from file through wolfSSL_X509_LOOKUP_load_file().
The CRL no longer has pass signature validation. The validation will occur when used.

### CODE REVIEW

File: src/crl.c [hunk 1]
```diff
@@ -89,6 +90,34 @@ static int InitCRL_Entry(CRL_Entry* crle, DecodedCRL* dcrl)
...
+        crle->tbsSz = dcrl->sigIndex - dcrl->certBegin;
+        crle->signatureSz = dcrl->sigLength;
+        crle->signatureOID = dcrl->signatureOID;
+        crle->toBeSigned = XMALLOC(crle->tbsSz, heap, DYNAMIC_TYPE_CRL_ENTRY);
+        if (crle->toBeSigned == NULL)
+            return -1;
+        crle->signature = XMALLOC(crle->signatureSz, heap,
+                                  DYNAMIC_TYPE_CRL_ENTRY);
+        if (crle->signature == NULL) {
+            XFREE(crle->signature, crl->heap, DYNAMIC_TYPE_CRL_ENTRY);
```
Comment (toddouska) [inline src/crl.c]: Shouldn't this be free crle->toBeSigned?

File: src/crl.c [hunk 2]
```diff
@@ -167,6 +200,46 @@ static int CheckCertCRLList(WOLFSSL_CRL* crl, DecodedCert* cert, int *pFoundEntr
...
+                ret = VerifyCRL_Signature(crle->toBeSigned, crle->tbsSz,
+                    crle->signature, crle->signatureSz, crle->signatureOID, ca);
+
+                if (wc_LockMutex(&crl->crlLock) != 0) {
+                    WOLFSSL_MSG("wc_LockMutex failed");
+                    return BAD_MUTEX_E;
+                }
+
+                if (ret == 0)
+                    crle->verified = 1;
```
Comment (toddouska) [inline src/crl.c]: crle may have been freed while the lock was released.  I feel like we need to make a deep copy of crle to do the verification and after getting the long again see if crle still exists.

File: src/crl.c (hunk 1 above)
Comment (SparkiDev) [inline src/crl.c]: Yep! Fixed.

File: src/crl.c (hunk 2 above)
Comment (SparkiDev) [inline src/crl.c]: Yeah, can't see anyway around it.
I now free the toBeSigned and signature part of the crle once verification attempted.

Comment (toddouska) [inline src/crl.c]: crle may no longer exist on line 259 after the unlock on line 233.  Don't we need to walk the list again and see if there's still an entry in it that matches before updating it?

Comment (SparkiDev) [inline src/crl.c]: Yep. Fixed to re-walk.

**Comment by kaleb-himes** [comment]
Just needs rebased with master and should be good to go. Windows testsuite issues have already been fixed in upstream.

File: src/crl.c [hunk 3]
```diff
@@ -167,6 +200,92 @@ static int CheckCertCRLList(WOLFSSL_CRL* crl, DecodedCert* cert, int *pFoundEntr
...
+                byte* tbs = NULL;
+                word32 tbsSz = crle->tbsSz;
+                byte* sig = NULL;
+                word32 sigSz = crle->signatureSz;
+                word32 sigOID = crle->signatureOID;
+                SignatureCtx sigCtx;
+
+                tbs = XMALLOC(tbsSz, crl->heap, DYNAMIC_TYPE_CRL_ENTRY);
+                if (tbs == NULL)
+                    return MEMORY_E;
```
Comment (toddouska) [inline src/crl.c]: return while crlLock is held seems like an error.

File: src/crl.c [hunk 4]
```diff
@@ -167,6 +200,92 @@ static int CheckCertCRLList(WOLFSSL_CRL* crl, DecodedCert* cert, int *pFoundEntr
...
+                sig = XMALLOC(sigSz, crl->heap, DYNAMIC_TYPE_CRL_ENTRY);
+                if (sig == NULL) {
+                    XFREE(tbs, crl->heap, DYNAMIC_TYPE_CRL_ENTRY);
+                    return MEMORY_E;
```
Comment (toddouska) [inline src/crl.c]: return while crlLock is held seems like an error.

File: src/crl.c (hunk 3 above)
Comment (SparkiDev) [inline src/crl.c]: Fixed

File: src/crl.c (hunk 4 above)
Comment (SparkiDev) [inline src/crl.c]: Fixed

**Pull request merged**